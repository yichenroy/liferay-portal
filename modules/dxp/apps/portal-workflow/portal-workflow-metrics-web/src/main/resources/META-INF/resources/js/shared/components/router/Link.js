import {jsonToUrl} from '../../util/url';
import React from 'react';

/**
 * Router link.
 * @extends React.Component
 */
export default class Link extends React.Component {
	render() {
		const {children, className, query = {}, text, title, to} = this.props;

		return (
			<a
				className={className}
				data-senna-off
				href={jsonToUrl({query, title, to})}
			>
				{text || children}
			</a>
		);
	}
}