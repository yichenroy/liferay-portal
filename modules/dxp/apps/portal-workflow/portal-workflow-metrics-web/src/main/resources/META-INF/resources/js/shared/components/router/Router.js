import 'custom-event-polyfill';
import controlMenuHeading from '../control-menu-heading/controlMenuHeading';
import {PAGE_CHANGE} from './RouterConstants';
import React from 'react';

/**
 * Router.
 * @extends React.Component
 */
export default class Router extends React.Component {
	constructor(props) {
		super(props);
		const url = String(document.location);

		this.state = {
			firstUrl: url,
			lastUrl: url,
			path: props.defaultPath,
			query: {}
		};
	}

	componentDidMount() {
		const onPageChanged = () => this.onPageChanged();

		window.addEventListener(PAGE_CHANGE, onPageChanged);

		this.unsub = () => {
			window.removeEventListener(PAGE_CHANGE, onPageChanged);
		};
	}

	componentWillUnmount() {
		this.unsub();
		this.unsub = null;
	}

	getQuery() {
		const url = String(document.location);

		return url
			.substring(url.lastIndexOf('#') + 1)
			.split('&')
			.filter(key => key !== '')
			.reduce((query, pairKey) => {
				const splitedKey = pairKey.split('=');

				query[splitedKey[0]] = splitedKey[1];

				return query;
			}, {});
	}

	onPageChanged() {
		const query = this.getQuery();

		const {_path: path} = query;
		let {_title: title} = query;

		const {defaultPath, paths} = this.props;
		const {lastUrl} = this.state;

		const isFirstPage = !path || path === defaultPath;

		const pathToFind = isFirstPage ? defaultPath : path;

		const componentRender = paths.find(router => router.path === pathToFind);

		if (componentRender) {
			if (!title) {
				({title} = componentRender);
			}

			if (isFirstPage) {
				controlMenuHeading({title});
			}
			else {
				controlMenuHeading({backPath: lastUrl, title});
			}
		}

		delete query['_path'];
		delete query['_title'];

		this.setState({
			lastUrl: String(document.location),
			path: pathToFind,
			query
		});
	}

	render() {
		const {paths} = this.props;
		const {path, query} = this.state;
		const componentRenderIndex = paths.findIndex(
			router => router.path === path
		);

		return (
			<div>
				{componentRenderIndex > -1 &&
					paths[componentRenderIndex].component(query)}
			</div>
		);
	}
}