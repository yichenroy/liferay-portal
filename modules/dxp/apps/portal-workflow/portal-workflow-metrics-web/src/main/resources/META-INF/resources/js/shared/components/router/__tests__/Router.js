import Link from '../Link';
import React from 'react';
import Router from '../Router';

beforeEach(() => {
	const vbody = document.createElement('div');
	vbody.className = 'vbody';
	vbody.innerHTML =
		'<ul class="control-menu-nav"><label class="control-menu-level-1-heading">title</label></ul><ul class="control-menu-nav"></ul>';
	document.body.appendChild(vbody);
});

afterEach(() => {
	const vbodyElement = document.body.getElementsByClassName('vbody')[0];
	vbodyElement.parentNode.removeChild(vbodyElement);
});

test('Should test component render', () => {
	const component = shallow(<Router defautPath="test" paths={[]} />);
	const instance = component.instance();
	instance.componentDidMount();
	instance.componentWillUnmount();

	const link = shallow(
		<Link className="test" text="test" to="test" type="button" />
	);
	link.find('a').simulate('click');

	expect(component).toMatchSnapshot();
});

test('Should test navigate', () => {
	const paths = [
		{
			component: ({title}) => <div>{title}</div>,
			path: 'test-1',
			title: 'test-1'
		},
		{
			component: () => <div>{'test-2'}</div>,
			path: 'test-2',
			title: 'test-2'
		},
		{
			component: ({title}) => <div>{title}</div>,
			path: 'test-3',
			title: 'test-3'
		}
	];
	const component = shallow(<Router defautPath="test-1" paths={paths} />);
	const instance = component.instance();
	instance.componentDidMount();
	instance.componentWillUnmount();

	const link = shallow(
		<Link
			className="test2"
			query={{title: 'test-2'}}
			text="test-2"
			to="test-2"
			type="button"
		/>
	);
	link.find('a').simulate('click');

	expect(component).toMatchSnapshot();
});

test('Should test navigate with title', () => {
	const paths = [
		{
			component: ({title}) => <div>{title}</div>,
			path: 'test-1',
			title: 'test-1'
		},
		{
			component: () => <div>{'test-2'}</div>,
			path: 'test-2',
			title: 'test-2'
		},
		{
			component: ({title}) => <div>{title}</div>,
			path: 'test-3',
			title: 'test-3'
		}
	];
	const component = shallow(<Router defautPath="test-3" paths={paths} />);
	const instance = component.instance();
	instance.componentDidMount();
	instance.componentWillUnmount();

	const link = shallow(
		<Link
			className="test2"
			query={{title: 'test-2'}}
			text="test"
			to="test-2"
			type="button"
		/>
	);
	link.find('a').simulate('click');

	expect(component).toMatchSnapshot();
});

test('Should test returning navigate', () => {
	const paths = [
		{
			component: ({title}) => <div>{title}</div>,
			path: 'test-1',
			title: 'test-1'
		},
		{
			component: () => <div>{'test-2'}</div>,
			path: 'test-2',
			title: 'test-2'
		},
		{
			component: ({title}) => <div>{title}</div>,
			path: 'test-3',
			title: 'test-3'
		}
	];
	const component = shallow(<Router defautPath="test-3" paths={paths} />);
	const instance = component.instance();
	instance.componentDidMount();
	instance.componentWillUnmount();

	const link = shallow(
		<Link
			className="test2"
			query={{title: 'test-2'}}
			text="test2"
			title="test-2"
			to="test-2"
			type="button"
		/>
	);
	link.find('a').simulate('click');

	const linkreturn = shallow(
		<Link
			className="test3"
			query={{title: 'test-3'}}
			text="test3"
			to="test-3"
		/>
	);
	linkreturn.find('a').simulate('click');

	expect(component).toMatchSnapshot();
});

test('Should test returning navigate by click', () => {
	const paths = [
		{
			component: () => <div>{'test-1'}</div>,
			path: 'test-1',
			title: 'test-1'
		},
		{
			component: () => <div>{'test-1'}</div>,
			path: 'test-2',
			title: 'test-2'
		}
	];
	const component = shallow(<Router defautPath="test-1" paths={paths} />);
	const instance = component.instance();
	instance.componentDidMount();
	instance.componentWillUnmount();

	const link = shallow(
		<Link
			className="test2"
			query={{title: 'test-2'}}
			text="test2"
			title="test-2"
			to="test-2"
			type="button"
		/>
	);
	link.find('a').simulate('click');

	const linkReturn = document.body.getElementsByClassName(
		'metric-control-menu-heading'
	)[0];
	linkReturn.dispatchEvent(new Event('click'));

	expect(component).toMatchSnapshot();
});