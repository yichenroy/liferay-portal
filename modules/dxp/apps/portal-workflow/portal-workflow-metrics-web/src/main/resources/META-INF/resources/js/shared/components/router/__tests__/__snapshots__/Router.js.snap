// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should navigate when event is dispatched 1`] = `
<div>
  <div>
    test-2
  </div>
</div>
`;

exports[`Should test component render 1`] = `<div />`;

exports[`Should test navigate 1`] = `
<div>
  <div />
</div>
`;

exports[`Should test navigate with title 1`] = `
<div>
  <div />
</div>
`;

exports[`Should test returning navigate 1`] = `
<div>
  <div />
</div>
`;

exports[`Should test returning navigate by click 1`] = `
<div>
  <div>
    test-1
  </div>
</div>
`;
